Autoscale frozen tier into existence

henningandersen opened the pull request:
This commit adds two related changes:
* ILM WaitForDataTierStep
* Autoscaling frozen_existence decider

The first part ensures that we wait mounting an index until a node that
can hold the index is available, avoiding a failed restore and red
cluster state. This is in particular important for the frozen phase, but
is done generically in the searchable snapshot action.

The second part triggers on indices in the ILM frozen phase to scale the
tier into existence by requiring a minimal amount of memory and storage.

Closes #72771

henningandersen committed:
Autoscale frozen tier into existence

This commit adds two related changes:
* ILM WaitForDataTierStep
* Autoscaling frozen_existence decider

The first part ensures that we wait mounting an index until a node that
can hold the index is available, avoiding a failed restore and red
cluster state. This is in particular important for the frozen phase, but
is done generically in the searchable snapshot action.

The second part triggers on indices in the ILM frozen phase to scale the
tier into existence by requiring a minimal amount of memory and storage.

Closes #72771

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

henningandersen committed:
Fix docs

henningandersen committed:
fix nit

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -116,7 +116,6 @@ public String getSnapshotRepository() {
         // here before generating snapshots that can't be used if the user doesn't have the right license level.
         BranchingStep conditionalSkipActionStep = new BranchingStep(preActionBranchingKey, checkNoWriteIndex, nextStepKey,
             (index, clusterState) -> {
-                XPackLicenseState licenseState = XPackPlugin.getSharedLicenseState();
```
Can you explain why we need to move the license state back up into the top level instead of getting it here? It added quite a few lines injecting it into every ILM action

henningandersen replied:
The problem is that `XPackPlugin.getSharedLicenseState()` is backed by a static `SetOnce` field making it unsuitable for use in tests. This pattern of passing in the `licenseState` was taken from ML. In CCR and Autoscaling, the `licenceState` is wrapped in a checker that is passed around instead. For ILM, the ML approach felt most natural since it allows any license check in the future.

I think there is no internal cluster tests today that trigger the `SearchableSnapshotAction`, however, I needed to add one in `FrozenExistenceDeciderIT`, necessitating this change.

If there is a different and/or better pattern for doing this I should be more than happy to adapt this.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -206,12 +205,17 @@ public String getSnapshotRepository() {
         // can be used, it jumps directly here, skipping the snapshot generation steps above.
+        WaitForDataTierStep waitForDataTierStep =
+            new WaitForDataTierStep(waitForDataTierKey, mountSnapshotKey,
+                MountSearchableSnapshotRequest.getDataTiersPreference(storageType));
```
This is technically going to be sending the wrong preference, because the preference is either `data_cold,data_warm,data_hot` or `data_frozen` from the `getDataTiersPreference`, but actually if this is the hot phase, it will end up with a different preference:

https://github.com/elastic/elasticsearch/blob/c5b9fdea5bf7ac5988f7dde71a61754a1a3b50fb/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java#L108-L111

henningandersen replied:
thanks for noticing, fixed in 3d23644c27086a84dcb88c79a99fc6644da675f8

dakrone reviewed (commented):
I left some initial comments about the ILM-specific stuff

henningandersen commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java:
```diff
@@ -187,7 +193,7 @@ protected Clock getClock() {
             clusterService, threadPool));
         indexLifecycleInitialisationService.set(new IndexLifecycleService(settings, client, clusterService, threadPool,
-            getClock(), System::currentTimeMillis, xContentRegistry, ilmHistoryStore.get()));
+            getClock(), System::currentTimeMillis, xContentRegistry, ilmHistoryStore.get(), getLicenseState()));
```
I believe not, since `XPackPlugin.licenseState` is a `SetOnce` object. Also, `XPackLicenseState` has an `update` method, which changes the `status` and is invoked from `LicenseService`. Must admit I am also not too familiar with the details, but I do believe this does not change behavior "in production".

henningandersen commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleExecutionState.java:
```diff
@@ -97,6 +97,15 @@ public static LifecycleExecutionState fromIndexMetadata(IndexMetadata indexMetad
+     * @return true if frozen phase, false otherwise.
+     */
+    public static boolean isFrozenPhase(IndexMetadata indexMetadata) {
+        Map<String, String> customData = indexMetadata.getCustomData(ILM_CUSTOM_METADATA_KEY);
+        return customData != null && TimeseriesLifecycleType.FROZEN_PHASE.equals(customData.get(PHASE));
```
Yeah, I was deliberately micro-tuning this a bit, since it will be called rather heavily from the autoscaling decider and I felt like skipping the various object creations and map lookups etc. made sense.

dakrone replied:
Okay, if the performance difference is important enough, we might want to add a comment to make sure it doesn't get code-golfed down to a less optimal version in the future

henningandersen replied:
++, added in bfc5fa4bff9b0d0057a265bafbcc764d13e09e3d

henningandersen committed:
Wait for hot tier too

Now wait for hot tier in hot phase rather than any of cold,warm,hot.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

henningandersen committed:
Add comment to avoid de-tuning.

henningandersen committed:
Use SingleMessageFieldInfo

henningandersen commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForDataTierStep.java:
```diff
@@ -0,0 +1,101 @@
+        return Objects.hash(super.hashCode(), tierPreference);
+    }
+
+    static final class Info implements ToXContentObject {
```
++, fixed in 84b99ef3e8ab7a4f91fc8be95238c4a9fe9b65d8

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

henningandersen committed:
Smaller impact change

By adding a default method to LifecycleAction we can reduce the impact
on every step in ILM.

henningandersen committed:
Rollback a few more nit changes.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

henningandersen committed:
Cleanup

henningandersen committed:
cs

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

dakrone reviewed (approved):
ILM parts LGTM, thanks for iterating on this Henning

DaveCTurner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -92,6 +91,11 @@ public String getSnapshotRepository() {
 
     @Override
     public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
+        throw new UnsupportedOperationException();
```
nit: `assert false` here too? For the record I'd rather just add the parameter to `toSteps()` everywhere rather than have this uncalled overload, i.e, the opposite of what we did in d4b2e32137b429275fc1372630fc33ce49bb3887.

henningandersen replied:
See 63816a07d4ab7802b2be8b0721d99d1b31df1c3d. I left the overload as is.

DaveCTurner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/searchablesnapshots/MountSearchableSnapshotRequest.java:
```diff
@@ -109,6 +110,22 @@ public MountSearchableSnapshotRequest(StreamInput in) throws IOException {
+     * performing a full mount the preference is cold - warm - hot. When
+     * performing a partial mount the preference is only frozen
+     */
+    public static String getDataTiersPreference(Storage type) {
```
nit/qq: why not an instance method on `Storage`?

henningandersen replied:
Doh, thanks, fixed in 32371ac45b6220d63bcaafdedc9b38f6ca4d5440

DaveCTurner commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotActionTests.java:
```diff
@@ -47,15 +48,29 @@ public void testToSteps() {
         assertThat(steps.get(12).getKey(), is(expectedSteps.get(12)));
         assertThat(steps.get(13).getKey(), is(expectedSteps.get(13)));
         assertThat(steps.get(14).getKey(), is(expectedSteps.get(14)));
+        assertThat(steps.get(15).getKey(), is(expectedSteps.get(15)));
```
nit: what's the deal with all this unrolling?

henningandersen replied:
I mainly followed the same methodology. I will leave it to @dakrone to comment on the idea behind this?

dakrone replied:
There's no reason these have to be unrolled, as long as we have the expected steps comparing them all at once is fine. I don't know why they were initially unrolled.

DaveCTurner reviewed (approved):
LGTM I think, sorry it's taken a while to get my head around this. I left only trivial and optional comments.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_wait_for_frozen_tier

henningandersen committed:
post merge fixes

henningandersen committed:
Storage.defaultDataTiersPreference

henningandersen committed:
Proper javadoc

henningandersen committed:
assert false

henningandersen commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecyclePolicy.java:
```diff
@@ -206,9 +207,10 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      *
      * @param client The Elasticsearch Client to use during execution of {@link AsyncActionStep}
      *               and {@link AsyncWaitStep} steps.
+     * @param licenseState
```
I added a bit of doc in aec84ec87d84c33389d8883db3db61261ffb6cc7

henningandersen committed:
nits
